tests/app/celery/test_reporting_tasks.py

```diff
@@ -61,39 +61,58 @@ def test_create_nightly_billing_triggers_tasks_for_days(notify_api, mocker, day_
         assert mock_celery.apply_async.call_args_list[i][1]['kwargs'] == {'process_day': expected_kwargs[i]}
 
 
-@freeze_time('2019-08-01')
-def test_create_nightly_notification_status_triggers_tasks(notify_api, sample_service, mocker):
-    mock_celery = mocker.patch('app.celery.reporting_tasks.create_nightly_notification_status_for_service_and_day')
+@freeze_time('2019-08-01T00:30')
+def test_create_nightly_notification_status_triggers_tasks(
+    notify_api,
+    sample_service,
+    sample_template,
+    mocker,
+):
+    mock_celery = mocker.patch(
+        'app.celery.reporting_tasks.create_nightly_notification_status_for_service_and_day'
+    ).apply_async
+
+    create_notification(template=sample_template, created_at='2019-07-31')
     create_nightly_notification_status()
 
-    assert mock_celery.apply_async.call_count == (
-        (4 * 3)  # four days, three notification types
-        +
-        6  # six more days of just letters
+    mock_celery.assert_called_with(
+        kwargs={
+            'service_id': sample_service.id,
+            'process_day': '2019-07-31',
+            'notification_type': SMS_TYPE
+        },
+        queue=QueueNames.REPORTING
     )
 
-    for process_date, notification_type in itertools.product(
-        ['2019-07-31', '2019-07-30', '2019-07-29', '2019-07-28'],
-        [SMS_TYPE, EMAIL_TYPE, LETTER_TYPE]
-    ):
-        mock_celery.apply_async.assert_any_call(
-            kwargs={
-                'process_day': process_date,
-                'notification_type': notification_type,
-                'service_id': sample_service.id,
-            },
-            queue=QueueNames.REPORTING
-        )
 
-    for process_date in ['2019-07-27', '2019-07-26', '2019-07-25', '2019-07-24', '2019-07-23', '2019-07-22']:
-        mock_celery.apply_async.assert_any_call(
-            kwargs={
-                'process_day': process_date,
-                'notification_type': LETTER_TYPE,
-                'service_id': sample_service.id,
-            },
-            queue=QueueNames.REPORTING
-        )
+@freeze_time('2019-08-01T00:30')
+@pytest.mark.parametrize('notification_date, expected_types_aggregated', [
+    ('2019-08-01', set()),
+    ('2019-07-31', {EMAIL_TYPE, SMS_TYPE, LETTER_TYPE}),
+    ('2019-07-28', {EMAIL_TYPE, SMS_TYPE, LETTER_TYPE}),
+    ('2019-07-27', {LETTER_TYPE}),
+    ('2019-07-22', {LETTER_TYPE}),
+    ('2019-07-21', set()),
+])
+def test_create_nightly_notification_status_triggers_relevant_tasks(
```

Comment:
can we add a test where a service _isn't_ picked up when it has notifications on a different day?

it's probably easier to write that test forthe new `get_service_ids_with_notifications_on_date` function in the equivalent dao test file.

but i'm also interested to see us add some BST/GMT tests on this `create_nightly_notification_status` task where we ensure it sends the correct timestamps through to the `get_service_ids_with_notifications_on_date`

Reply:
This is implicit in the first and last scenarios: a notification exists for a service, but nothing happens. I suppose there's a weird potential bug where we call the subtask for the full range of days if it has a notification on just one of them. That's covered by the previous test, though.

One thing I haven't done is check this works for multiple services when they _both_ have relevant notifications. I tend to avoid testing "is it a loop" though, on the basis that it's hard to write a non-looping bit of code that only runs once vs. one that can work for any number of items => a low value test.

Ramble over. Do you think there's still a missing test here?

Reply:
i think what i was concerned about was that i'd spotted a bug, and we didn't have any tests that covered that. because your test  was only concerned with the outer python loops to ensure that the right notification types are invoked for the right day

get_service_ids_with_notifications_on_date now has tests (added in a 966c4db8c) which I think is what I was hoping for 👍